File: packages/fsm-core-async-op-worker/src/worker-sdk/typescript/validate-async-operation.ts
// Discovers and verifies typescript promise-actor folders, mirroring
// packages/fsm-compiler-ts/src/validate-async-operation-logic.ts's
// validateAsyncOperationFromFolders — trimmed to typescript only: no
// `runtimeLanguages` argument, `lang` fixed to "typescript" instead of
// looped over, and no python/go/rust branches (the per-language `if`
// dropped along with them, since there's only one language left to check).
//
// The typescript check also skips the subprocess hop: the real compiler's
// validator shells out to `check_fn.ts` via `Deno.Command` because it must
// support checkers that aren't Deno/TS (python3, a compiled go/rust binary).
// This file only ever checks typescript, in a process that's already Deno —
// so it inlines check_fn.ts's own check (dynamic `import()` + `typeof
// mod[fnName] === "function"`) directly, in-process.

import { getLogger } from "@logtape/logtape";
import {
  type ActorPluginValidationResult,
  type ActorReference,
  isVersionFolderName,
  type OperationLang,
  type WorkflowType,
} from "@pgfsm/compiler";

const logger = getLogger([
  "@pgfsm/worker",
  "async-op-worker-gateway",
  "worker-sdk-ts",
  "validate",
]);

const lang: OperationLang = "typescript";

export async function validateAsyncOperationFromFoldersTypescript(
  folderPath: string,
  _workflowType: WorkflowType,
  skipDirs: string[] = [],
  _availableActors: ActorReference[] = [],
): Promise<ActorPluginValidationResult[]> {
  if (folderPath.startsWith(".")) {
    throw new Error(
      `Invalid folder path: ${folderPath}. Folder paths cannot start with '.'`,
    );
  }
  if (folderPath.endsWith("/")) {
    throw new Error(
      `Invalid folder path: ${folderPath}. Folder paths cannot end with '/'`,
    );
  }
  if (folderPath.startsWith("/")) {
    logger.info("Importing workflows from absolute path: {path}", {
      path: folderPath,
    });
  } else {
    logger.info("Importing workflows from relative path: {path} to {cwd}", {
      path: folderPath,
      cwd: Deno.cwd(),
    });
  }
  const absFolderPath = folderPath.startsWith("/")
    ? folderPath
    : `${Deno.cwd()}/${folderPath}`;

  const allFolderResults: ActorPluginValidationResult[] = [];
  try {
    const stat = await Deno.stat(absFolderPath);
    if (!stat.isDirectory) {
      logger.error("Provided path is not a directory: {path}", {
        path: absFolderPath,
      });
      throw new Error(`Provided path '${absFolderPath}' is not a directory.`);
    }
    for await (const dirEntry of Deno.readDir(absFolderPath)) {
      if (dirEntry.isDirectory) {
        if (skipDirs.includes(dirEntry.name)) continue;

        const fsmDirPath = `${absFolderPath}/${dirEntry.name}`;

        for await (const subEntry of Deno.readDir(fsmDirPath)) {
          if (subEntry.isDirectory) {
            if (isVersionFolderName(subEntry.name)) {
              const absVersionPath = `${fsmDirPath}/${subEntry.name}`;
              try {
                const langPath = `${absVersionPath}/${lang}`;
                try {
                  await Deno.stat(langPath);
                } catch {
                  logger.info(
                    "Lang folder {lang} not found in {path}, skipping",
                    { lang, path: absVersionPath },
                  );
                  continue;
                }

                const actorsPath = `${langPath}/actors`;
                try {
                  await Deno.stat(actorsPath);
                } catch {
                  logger.info(
                    "Actors folder not found for {lang} in {path}, skipping",
                    { lang, path: langPath },
                  );
                  continue;
                }

                for await (const actorDir of Deno.readDir(actorsPath)) {
                  if (!actorDir.isDirectory) continue;

                  const fnName = actorDir.name;
                  const modulePath = `${actorsPath}/${fnName}/${fnName}.ts`;

                  try {
                    await Deno.stat(modulePath);
                  } catch {
                    logger.warn(
                      "actors: expected file not found for actor {actor} in {path}, skipping",
                      { actor: fnName, path: modulePath },
                    );
                    continue;
                  }

                  let isVerified = false;
                  let errorMessage: string | null = null;

                  try {
                    const mod = await import(`file://${modulePath}`);
                    if (typeof mod[fnName] === "function") {
                      isVerified = true;
                      logger.info(
                        "actors/typescript: function {src} found in {path}",
                        { src: fnName, path: modulePath },
                      );
                    } else {
                      errorMessage =
                        `'${fnName}' is not exported as a function from ${modulePath}`;
                      logger.error(
                        "actors/typescript: function {src} not found in {path}: {err}",
                        { src: fnName, path: modulePath, err: errorMessage },
                      );
                    }
                  } catch (err) {
                    errorMessage = `Failed to import ${modulePath}: ${err}`;
                    logger.error(
                      "actors/typescript: function {src} not found in {path}: {err}",
                      { src: fnName, path: modulePath, err: errorMessage },
                    );
                  }

                  allFolderResults.push({
                    src: fnName,
                    method: fnName,
                    fsmName: fnName,
                    fsmType: "promise",
                    fsmVersion: subEntry.name,
                    fsmLanguage: lang,
                    isVerified,
                    fsmModulePath: modulePath,
                    parentFsmName: dirEntry.name,
                    parentFsmVersion: subEntry.name,
                    comment:
                      "for fsmType promise fsmVersion will be its parentFsmVersion value",
                    parentFsmPath: fsmDirPath,
                    errorMessage,
                  });
                }
              } catch (err) {
                logger.error("Failed to validate {path}: {error}", {
                  path: absVersionPath,
                  error: err,
                });
              }
            } else {
              logger.info(
                "Skipping non-versioned folder: {name} in {dir}",
                { name: subEntry.name, dir: fsmDirPath },
              );
            }
          }
        }
      }
    }
  } catch (err) {
    logger.error(
      "Error occurred while reading directory {path}: {error}",
      { path: absFolderPath, error: err },
    );
  }

  return allFolderResults;
}

File: packages/fsm-core-async-op-worker/src/worker-sdk/go/validate_async_operation.go
// Discovers and verifies go promise-actor folders.
//
// Go counterpart of ../typescript/validate-async-operation.ts's
// validateAsyncOperationFromFoldersTypescript, walking
// <folderPath>/<fsmName>/<version>/go/actors/<actorName>/<actorName>.go
// (this repo's real FSM actor convention — see
// apps/fsm-core-example/fsm/creditCheck/v01/go/actors/checkReportsTable/checkReportsTable.go)
// instead of the typescript file's .../typescript/actors/... path.
//
// Verification inlines packages/fsm-compiler-ts/src/checkers/check_fn.go's
// own check (go/ast walk for a FuncDecl matching the actor name, regardless
// of exported/unexported case — check_fn.go doesn't care) directly
// in-process, the same way the typescript/python/rust files inline their
// own checkers instead of shelling out.
//
// This is validation only, not invocation: unlike Python/TypeScript, a
// verified result here does not carry a callable — Go has no runtime
// mechanism to load a function out of a .go file (see sdk.go's package
// doc). The reference binary (main.go) still has to match verified results
// against a small compile-time registry of actually-linked-in handler
// functions — and unlike Rust, Go additionally enforces exports at compile
// time, so an unexported real-world actor function (like
// checkReportsTable, matching this repo's established stub convention)
// cannot be imported/linked at all without renaming it, which would break
// check_fn.go's own validation elsewhere in the repo. See main.go's
// knownHandler for the consequence.
package main

import (
	"go/ast"
	"go/parser"
	"go/token"
	"os"
	"path/filepath"
	"regexp"
	"sort"
)

const lang = "go"

// Mirrors @pgfsm/compiler's util.ts isVersionFolderName: /^v\d{2}$/.
var versionFolderRe = regexp.MustCompile(`^v\d{2}$`)

func hasFn(path string, fnName string) (bool, error) {
	fset := token.NewFileSet()
	f, err := parser.ParseFile(fset, path, nil, 0)
	if err != nil {
		return false, err
	}
	for _, decl := range f.Decls {
		if fn, ok := decl.(*ast.FuncDecl); ok && fn.Name.Name == fnName {
			return true, nil
		}
	}
	return false, nil
}

// Mirrors @pgfsm/compiler's ActorPluginValidationResult (util.ts).
type ActorPluginValidationResult struct {
	Src              string
	Method           string
	FsmName          string
	FsmType          string
	FsmVersion       string
	FsmLanguage      string
	IsVerified       bool
	FsmModulePath    string
	ParentFsmName    string
	ParentFsmVersion string
	ParentFsmPath    string
	ErrorMessage     string
}

func sortedDirEntries(dir string) []os.DirEntry {
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil
	}
	sort.Slice(entries, func(i, j int) bool { return entries[i].Name() < entries[j].Name() })
	return entries
}

func ValidateAsyncOperationFromFoldersGo(folderPath string, skipDirs []string) []ActorPluginValidationResult {
	var results []ActorPluginValidationResult

	skip := make(map[string]bool, len(skipDirs))
	for _, d := range skipDirs {
		skip[d] = true
	}

	info, err := os.Stat(folderPath)
	if err != nil || !info.IsDir() {
		return results
	}

	for _, fsmEntry := range sortedDirEntries(folderPath) {
		fsmName := fsmEntry.Name()
		if skip[fsmName] || !fsmEntry.IsDir() {
			continue
		}
		fsmDirPath := filepath.Join(folderPath, fsmName)

		for _, versionEntry := range sortedDirEntries(fsmDirPath) {
			version := versionEntry.Name()
			if !versionEntry.IsDir() || !versionFolderRe.MatchString(version) {
				continue
			}
			versionPath := filepath.Join(fsmDirPath, version)

			langPath := filepath.Join(versionPath, lang)
			if info, err := os.Stat(langPath); err != nil || !info.IsDir() {
				continue
			}

			actorsPath := filepath.Join(langPath, "actors")
			if info, err := os.Stat(actorsPath); err != nil || !info.IsDir() {
				continue
			}

			for _, actorEntry := range sortedDirEntries(actorsPath) {
				actorName := actorEntry.Name()
				if !actorEntry.IsDir() {
					continue
				}
				actorDir := filepath.Join(actorsPath, actorName)
				modulePath := filepath.Join(actorDir, actorName+".go")

				if info, err := os.Stat(modulePath); err != nil || info.IsDir() {
					continue
				}

				isVerified := false
				errorMessage := ""
				found, err := hasFn(modulePath, actorName)
				if err != nil {
					errorMessage = "Failed to parse " + modulePath + ": " + err.Error()
				} else if found {
					isVerified = true
				} else {
					errorMessage = "Function '" + actorName + "' not found in " + modulePath
				}

				results = append(results, ActorPluginValidationResult{
					Src:              actorName,
					Method:           actorName,
					FsmName:          actorName,
					FsmType:          "promise",
					FsmVersion:       version,
					FsmLanguage:      lang,
					IsVerified:       isVerified,
					FsmModulePath:    modulePath,
					ParentFsmName:    fsmName,
					ParentFsmVersion: version,
					ParentFsmPath:    fsmDirPath,
					ErrorMessage:     errorMessage,
				})
			}
		}
	}

	return results
}
